sources/api/apiclient/src/exec.rs

```diff
@@ -0,0 +1,683 @@
+    // not to send other types of messages first or the server won't have a process to act on and
+    // will reject us.  It'd be nice to send initialization parameters in the HTTP request body,
+    // but not all WebSocket clients support it.)
+    debug!(
+        "Sending initialize request for target '{}' with tty: {} and command: {:?}",
+        target,
+        terminal.tty().is_some(),
+        command
+    );
+    let init = Initialize {
+        command,
+        target,
+        tty: terminal.tty().clone(),
+    };
+    // Control messages go to the server in a text channel, so we serialize to JSON before sending.
+    let msg = serde_json::to_string(&ClientMessage::Initialize(init)).context(error::Serialize)?;
+    ws_tx
+        .unbounded_send(Message::Text(msg))
+        .context(error::SendMessage {
+            kind: "initialization",
+        })?;
+
+    // Now that the server knows what we want, we set up helper threads to manage communication.
+    // First, a heartbeat type that regularly pings the server and keeps track of responses.
+    let mut heartbeat = Heartbeat::new(ws_tx.clone());
+    // Next, a type that watches for signals to the local process, and either forwards them to the
+    // server (e.g. if you change your window size) or ends communication (e.g. for SIGTERM).
+    let mut signal_handler = HandleSignals::new(ws_tx.clone())?;
+
+    // We don't want to overload the server with our process input.  It sends us capacity updates
+    // to let us know how many more messages we can send before we should wait.  We keep track of
+    // that capacity in an AtomicCapacity that we can share across threads - the one where we
+    // receive messages from the server, and the one that's reading input to send to the server.
+    let capacity = Arc::new(AtomicCapacity::default());
+    let capacity_reader = Arc::clone(&capacity);
+
+    // Start a thread that reads input from the user and sends it across the WebSocket, waiting for
+    // capacity between reads if necessary.
+    let _ = ReadFromUser::new(ws_tx.clone(), capacity_reader, terminal.tty().is_some());
+    // Start a future that reads the stream of messages from the server.
+    let mut read_from_server = ReadFromServer::new(read, heartbeat.setter, capacity);
+
+    // We're all set up!  Wait for something that indicates we're done.
+    debug!("Waiting for completion: server, signal, or heartbeat");
+    let mut signal_ret = None;
+    tokio::select! {
+        res = &mut read_from_server.future => {
+            debug!("Server read completed");
+            // If our ReadFromServer thread hit an error, log it, except for the special case of a
+            // Close error, which is just an empty marker that we're done.
+            if let Err(e) = res {
+                let msg = e.to_string();
+                if !msg.is_empty() {
+                    error!("{}", e);
+                }
+            }
+        }
```

Comment:
We can only drop `terminal` once the program the user requested is complete, otherwise we'll screw up its input and output.  Any errors that we actually print before that point are taken to be serious enough that printing them, even in screwed up form, is worthwhile.  ("screwed up" here usually just means they'll appear in the middle of a line, since we don't print carriage returns by default, but in raw mode carriage returns aren't assumed.)

I'll clarify the comment on Terminal.

Reply:
Also, for this specific error case, I'll see if I can drop terminal before the message, since we know this is a "terminal" case.  (For the earlier debug messages, I'm not sure how we could make those appear normal; terminal settings apply to the whole terminal, not just stdout or stderr.  If you're in a TUI application, it seems like there wouldn't be a way to display anything without breaking its display, unless you somehow buffer all messages until after, or you log to a file.  But if you're in debug mode, you probably want to see messages even if it messes up your output a bit..?)